pkg/actions/service.go

```diff
@@ -357,22 +372,47 @@ func (s *StoreService) runTasks(ctx context.Context, record graveler.HookRecord,
 						task.HookRunID, task.Action.Name, task.HookID, task.Err)
 				}
 
-				err := hookOutputWriter.OutputWrite(ctx, &buf, int64(buf.Len()))
+				err = hookOutputWriter.OutputWrite(ctx, &buf, int64(buf.Len()))
 				if err != nil {
+					// non task error - we stop the tasks processing
 					return fmt.Errorf("failed to write action log. Run id '%s' action '%s' hook '%s': %w",
 						task.HookRunID, task.Action.Name, task.HookID, err)
 				}
+
+				// stop execution if needed - keep using lastErr for return value.
+				// we do not return here as we like to process all the steps
 				if task.Err != nil {
-					// stop execution of tasks and return error
-					return task.Err
+					// capture the first error
+					if actionErr == nil {
+						actionErr = task.Err
+					}
 				}
 			}
-			return nil
+			return actionErr
 		})
 	}
 	return g.Wait().ErrorOrNil()
 }
 
+func runHookIfEval(task *Task, actionErr error) (bool, error) {
+	result := actionErr == nil // by default, we run based on current status - success: no error
+	if task.If == "" {
+		return result, nil
+	}
+	env := map[string]any{
+		"success": func() bool { return actionErr == nil },
+		"failure": func() bool { return actionErr != nil },
+	}
+	output, err := expr.Eval(task.If, env)
```

Comment:
What limits `task.If`? How do protect from bad code execution?

Reply:
The expression evaluation language definition is found here: https://expr.medv.io/docs/Language-Definition.
It is built to evaluate expression no functionality that communicate or have side effect outside the expression itself.